doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,1015 @@
+   will 'capture' the entire hybrid workflow with **placeholder variables of
+   unknown value** used as the function arguments
+   (the **runtime arguments**).
+
+   These symbolic tracer objects represent **dynamic variable**, and are used
+   to determine how the JIT compiled function transforms its inputs to
+   outputs.
+
+#. **Compilation:** The captured program is then compiled to a parametrized
+   binary using the Catalyst compiler.
+
+#. **Execution:** Finally, the compiled function is executed with the
+   provided numerical function inputs, and the results returned.
+
+Once the function is first compiled, subsequent executions of the function
+will simply re-use the previous compiled binary, allowing steps (1) and (2) to
+be skipped. (Note: some cases, such as if the function argument types change,
+may trigger re-compilation.)
+
+For example, consider the following, where we print out a variable in the middle of
+our :func:`@qjit <~.qjit>` compiled function:
+
+>>> @qjit
+... def f(x):
+...     print(f"x = {x}")
+...     return x ** 2
+>>> f(2.)
+x = Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
+array(4.)
+>>> f(3.)
+array(9.)
+
+We can see that on the first execution, program capture/tracing occurs, and we
+can see the dynamic variable is printed (tracers capture *type*
+and *shape*, but not numeric value). This captured program is compiled, and
+then the binary is executed directly to return the function value --- the
+print statement is never invoked with the numerical value of ``x``.
+
+When we execute the function again, steps (1) and (2) are skipped since we
+have already compiled a binary; this is called directly to get the function
+result, and again the print statement is never hit.
+
+This allows us to distinguish between computations that happen
+at **compile-time** (steps 1 and 2), such as the ``print`` statement above,
+and those that happen at **runtime** (step 3).
+
+.. note::
+
+    As a general rule of thumb, for a function that is repeatedly executed
+    with different parameters, we want as much evaluation as possible to
+    happen at compile time.
+
+    However, computations at compile time cannot depend on the value of
+    dynamic variable, since this is not known yet. It can only depend
+    on **static variables**, those whose values are known.
+
+.. note::
+
+    A general guideline when working with JIT compilation and Catalyst:
+
+    - Python control flow and third party libraries like NumPy and SciPy will
+      be evaluated at compile-time, and can only accept static variables.
+
+    - JAX functions, such as ``jax.numpy``, and Catalyst functions like
+      :func:`~.cond` and :func:`~.for_loop` will be evaluated at
+      runtime, and can accept dynamic variables.
+
+    Note that if AutoGraph is enabled, Catalyst will attempt to convert Python
+    control flow to its Catalyst equivalent to support dynamic variables.
```

Comment:
I first suggested adding the TF and Catalyst import ordering issue here in this doc page, but I think this should be added as a warning to the AutoGraph guide page as the AutoGraph support is still an experimental support and recalled in the doc in "note" blocks. What do you think @josh146 ?

Reply:
I agree :)